Stringification of all ops internally

shen-tian opened the pull request:
This follows on discussions from, and closes #138.

- Updates the handful of `:op` as keywords in the main body of the code.
- Updates most of the tests to use string ops. It's telling that most of what needed changing was the test suite. Also modified the test suite to be stricter on EDN transports (we'll need to update fastlane)
- The EDN transport is now entirely transparent! `edn/read` and `str` are sufficient.
- Added a `liberally-accept` wrapper to incoming messages, so we still accept keyword ops.
- Updated all documentation where I could find it to be consistent.
- Added the missing `uri-scheme`, left out of the original PR.
- Expanded the documentation section on EDN transport.

---

Before submitting a PR make sure the following things have been done:

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [x] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

Thanks!

*If you're just starting out to hack on nREPL you might find this [section of its
manual][1] extremely useful.*

[1]: https:/nrepl.org/nrepl/hacking_on_nrepl.html

bbatsov commented on doc/modules/ROOT/pages/ops.adoc:
```diff
@@ -4,11 +4,11 @@ This file is _generated_ by #'nrepl.impl.docs/-main
 == Operations
 
-=== `:clone`
+=== `"clone"`
```
I think here we shouldn't print them with quotes, as that looks a bit weird. `clone` would be fine.

shen-tian replied:
Done.

bbatsov commented on doc/modules/ROOT/pages/design/transports.adoc:
```diff
@@ -91,6 +91,8 @@ $ clj -R:nrepl -m nrepl.cmdline -t nrepl.transport/edn
+
+This may be useful in a couple of usecases: In some clients, including ClojureScript ones, where EDN is easier to handle than bencode. Furthermore, because this transport exposes more of the richer data types in the internals of nREPL and its middlewares, it may support more complex usecases.
+
+The introduction of the EDN transport does require more specification of the message format, whereby some vagueness was allowed for by the limitatioms of bencode. For example, `:op` values are strings, and `:status` is either a keyword, or a set of keywords. This may cause some surprised, unintended consequences. Thus tweaks to the EDN message format may be possible until the transport is declare ...
```
Probably we should add a couple of code examples here and in the bencode section, so it'd be easier for people to understand the differences.

shen-tian replied:
I've added more details to the section now. Let me know if it's in the right direction? I feel the next step in documentation is to actually write a section on *messages*. We can probably more formally adopt the convention that in documentation, we express messages in EDN, and the messages we send/receive over the wire is either whatever `bencode/write-bencode` produces or the EDN as given.

bbatsov replied:
> I've added more details to the section now. Let me know if it's in the right direction? I feel the next step in documentation is to actually write a section on messages. We can probably more formally adopt the convention that in documentation, we express messages in EDN, and the messages we send/receive over the wire is either whatever bencode/write-bencode produces or the EDN as given.

You're definitely moving in the right direction with the latest round of updates. I like your idea. 

> I've done most of a spec for messages, which is sitting in another branch. I suspect that would serve as message specs more effectively than written documentation, given the intended audience of tool authors.

 I generally see the two approaches as complimentary, as it's kind of weird to just point people to some specs without providing them any explanations. But I agree that for people working a lot with nREPL probably the specs are going to be the single source of truth.

shen-tian replied:
I've updated the Message section to refer to the transport section for details, and elaborated a bit more on how the bencode transport works.

bbatsov commented on src/clojure/nrepl/server.clj:
```diff
@@ -19,11 +19,19 @@
+(defn- liberally-accept
+  "Accept messages that are not quite in spec. This comes into effect with
+   The EDN transport, and other transports that allow more types/data structures
+   than bencode, as there's more oppertunity to be out of specification."
```
opportunity

shen-tian replied:
Fixed.

bbatsov commented on src/clojure/nrepl/server.clj:
```diff
@@ -19,11 +19,19 @@
     (catch Throwable t
       (log t "Unhandled REPL handler exception processing message" msg))))
 
+(defn- liberally-accept
```
I think this name is a bit misleading, as the function doesn't really accept anything - it just normalizes the message map. Perhaps something like `normalize-msg` or whatever would be a better name?

shen-tian replied:
Let's go with `normalize-msg`. I was alluding to Postel's law with "liberally accept", but normalize is probably more accurate.

bbatsov commented:
Apart from my small feedback I just added, I think this PR is pretty much ready for merging in my opinion.

shen-tian commented:
> Apart from my small feedback I just added, I think this PR is pretty much ready for merging in my opinion.

I added a bit more documentation since your review. Am happy with where this is now.

bbatsov commented:
Thanks! Great work, overall!

Btw, I wonder if we should also add something about dealing with Boolean and float/double in the docs. And the encoding of nil. Probably those are some of the most common points of confusions for newcomers.

shen-tian commented:
> Thanks! Great work, overall!
> 
> Btw, I wonder if we should also add something about dealing with Boolean and float/double in the docs. And the encoding of nil. Probably those are some of the most common points of confusions for newcomers.

Hmm... haven't even thought how those worked. `nil` encodes to empty array, right?

bbatsov commented:
I think so. But in practice we don't use booleans and nils anywhere due to their fuzzy semantics.
